kasa/cleartexttransport.py

```diff
@@ -0,0 +1,241 @@
+
+_LOGGER = logging.getLogger(__name__)
+
+
+ONE_DAY_SECONDS = 86400
+SESSION_EXPIRE_BUFFER_SECONDS = 60 * 20
+
+
+def _md5(payload: bytes) -> str:
+    algo = hashlib.md5()  # noqa: S324
+    algo.update(payload)
+    return algo.hexdigest()
+
+
+class TransportState(Enum):
+    """Enum for transport state."""
+
+    LOGIN_REQUIRED = auto()  # Login needed
+    ESTABLISHED = auto()  # Ready to send requests
+
+
+class CleartextTransport(BaseTransport):
+    """Implementation of the cleartext transport protocol.
+
+    This transport uses HTTPS without any further payload encryption.
+    """
+
+    DEFAULT_PORT: int = 4433
+    COMMON_HEADERS = {
+        "Content-Type": "application/json",
+    }
+    BACKOFF_SECONDS_AFTER_LOGIN_ERROR = 1
+
+    def __init__(
+        self,
+        *,
+        config: DeviceConfig,
+    ) -> None:
+        super().__init__(config=config)
+
+        if (
+            not self._credentials or self._credentials.username is None
+        ) and not self._credentials_hash:
+            self._credentials = Credentials()
+        if self._credentials:
+            self._login_params = self._get_login_params(self._credentials)
+        else:
+            # TODO: Figure out how to handle credential hash
+            self._login_params = json_loads(
+                base64.b64decode(self._credentials_hash.encode()).decode()  # type: ignore[union-attr]
+            )
+        self._default_credentials: Credentials | None = None
+        self._http_client: HttpClient = HttpClient(config)
+
+        self._state = TransportState.LOGIN_REQUIRED
+        self._session_expire_at: float | None = None
+
+        self._app_url = URL(f"https://{self._host}:{self._port}/app")
+        self._token_url: URL | None = None
+
+        _LOGGER.debug("Created cleartext transport for %s", self._host)
+
+    @property
+    def default_port(self) -> int:
+        """Default port for the transport."""
+        return self.DEFAULT_PORT
+
+    @property
+    def credentials_hash(self) -> str:
+        """The hashed credentials used by the transport."""
+        return base64.b64encode(json_dumps(self._login_params).encode()).decode()
+
+    def _get_login_params(self, credentials: Credentials) -> dict[str, str]:
+        """Get the login parameters based on the login_version."""
+        un, pw = self.hash_credentials(credentials)
+        # The password hash needs to be upper-case
+        return {"password": pw.upper(), "username": un}
+
+    @staticmethod
+    def hash_credentials(credentials: Credentials) -> tuple[str, str]:
+        """Hash the credentials."""
+        un = credentials.username
+        pw = _md5(credentials.password.encode())
+        return un, pw
+
+    def _handle_response_error_code(self, resp_dict: Any, msg: str) -> None:
+        """Handle response errors to request reauth etc.
+
+        TODO: This should probably be moved to the base class as
+         it's common for all smart protocols?
+        """
```

Comment:
The response error codes at the transport level are not common to all transports, i.e. the klap transport does not raise them.  I think this transport probably is a flavour of the `AES` transport in that it uses the same login messages and response error codes.  Perhaps this class should be named `AesSslTransport` and derive from `AesTransport` only overriding the `perform_login` and `send` methods (it's not really clear text as it is encrypted via SSL). `Supports HTTPS:     True` could be used by the protocol factory to decide whether to return  `AesSslTransport` or `AesTransport`.

Reply:
Extending from `AesTransport` sounds indeed like a good idea, but `perform_handshake()` would also require override to be a no-op. I'll take a look into that at some later point, as this was rather created as a PoC and is not needed for the 0.7.0 release.